### ISSUE TITLE
WithTimeout wont stop timed-out tasks

### ISSUE THREAD

**Issue by pcampbell83** [jamesmontemagno/mvvm-helpers#59]
Hi, looking at the implementation -- won't the net result be that; the task itself will not stop running even if the timeout succeeds first?

Is that intentional?

**Comment by mjfreelancing** [comment]
I agree. Even if the caller checks the return result against `default` this is no indication the task timed out because the running task may actually return `null` if T is a reference type. And yes, if the delay task completes first then the original task will remain running.

The only way to do this correctly is for the caller to use a `CancellationTokenSource` and have this cancelled if the delay task completes first. It's then up to the caller to handle this situation within their own logic by checking `IsCancellationRequested` on the `CancellationToken`.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/cancel-remaining-async-tasks-after-one-is-complete

https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken.iscancellationrequested?view=netframework-4.8
